### DISCUSSION TITLE
Quick unlock: PIN

### DISCUSSION THREAD

**Discussion by msladek** [hazcod/enpass-cli#110] [Ideas]
The desktop app offers multiple quick unlock options, see [here](https://www.enpass.io/docs/security-whitepaper-enpass/quick_unlock.html). It would also be convenient to have something similar for the cli. Biometrics might be complicated, but a PIN should be doable.

Now there are multiple ways to do this, but the desktop app "does not actually close the Enpass database but adds just an extra screen for authorization". I don't think this can be done easily without writing a daemon that keeps the db connection open.

We could however encrypt the [derived database key](https://github.com/hazcod/enpass-cli/blob/afdb4a9a5aee5750af78b13cb049d97a06ee8738/pkg/enpass/vault.go#L125) with the PIN and write it to (ram)disk. This file could be checked for consecutive runs and decrypted by only providing the PIN.

- this PIN feature would be disabled by default and only be prompted if a `-pin` flag is provided.
- require PIN min length, e.g. 8 characters
- encryption could be done similar to how the Enpass database itself is encrypted, e.g. with this [aes-256-gcm](https://gist.github.com/tscholl2/dc7dc15dc132ea70a98e8542fefffa28#file-aes-go)
- the encrypted file shouldn't survive a system restart, thus a ramdisk like `/run/user/$UID` aka `XDG_RUNTIME_DIR
` or `/dev/shm` or `/tmp` for darwin seem like suitable locations.
- `600` file permissions

@hazcod WDYT? Is this idea sensible?

**Comment by fasher** [comment]
I would love if an agent daemon can be implemented so we can login once (within a se time)  and keep a session env variable that can communicate with the agent that does the actual decryption. 
lastpass-cli has something like that implemented

**Reply by msladek** [reply]
while an agent daemon certainly has many advantages, it is also much more complex to implement and setup. I feel this would overwhelm the current development capacity of this project. but of course any help is welcome.

**Comment by hazcod** [comment]
@msladek That sounds like a reasonable implementation.

**Reply by msladek** [reply]
thanks for your assessment. you can expect a PR addressing this within a few days.
